src/librustc_typeck/error_codes.rs

```diff
@@ -2425,6 +2425,83 @@ struct Bar<S, T> { x: Foo<S, T> }
 ```
 "##,
 
+E0307: r##"
+This error indicates that the `self` parameter in a method has an invalid
+"reciever type".
+
+Methods take a special first parameter, of which there are three variants:
+`self`, `&self`, and `&mut self`. The type `Self` acts as an alias to the
+type of the current trait implementor, or "receiver type". Besides the
+already mentioned `Self`, `&Self` and `&mut Self` valid receiver types, the
+following are also valid, if less common: `self: Box<Self>`,
+`self: Rc<Self>`, `self: Arc<Self>`, and `self: Pin<P>` (where P is one of
+the previous types except `Self`).
+
+```
+# struct Foo;
+trait Trait {
+    fn foo(&self);
+//         ^^^^^ this let's you refer to the type that implements this trait
+}
+impl Trait for Foo {
+//             ^^^ this is the "receiver type"
+    fn foo(&self) {}
+//         ^^^^^ this is of type `Foo`
+}
+```
+
+The above is equivalent to:
+
+```
+# struct Foo;
+# trait Trait {
+#     fn foo(&self);
+# }
+impl Trait for Foo {
+    fn foo(&self: &Foo) {}
+}
+```
+
+When using an invalid reciver type, like in the following example,
+
+```compile_fail,E0307
+# struct Foo;
+# struct Bar;
+# trait Trait {
+#     fn foo(&self);
+# }
+impl Trait for Struct {
+    fn foo(&self: &Bar) {}
+}
+```
```

Comment:
You have said *"When P,"* but I'm waiting for *"then Q"*.
Is it the pharagraph below? That's not clear from the text.

Reply:
I forgot to finish the sentence because I didn't know how to end it ^_^